### PR TITLE
Rewrite CA3075 with IOperation analyzer

### PR DESCRIPTION
**Pull request by qinxgit**
Rewrite CA3075 with IOperation analyzer.

### CODE REVIEW

**Comment by qinxgit** [comment]
@genlu @srivatsn Rewrited CA3075 with IOperation analyzer.

**Comment by qinxgit** [comment]
@genlu @srivatsn Take a short time to give me a feedback on obvious/big issues. I plan to get check-in at the end of this sprint, which is before 3/4.

File: src/Desktop.Analyzers/Core/DesktopAnalyzersResources.resx [hunk 1]
```diff
@@ -368,7 +368,7 @@
     <comment>CA3075</comment>
   </data>
   <data name="DoNotUseDtdProcessingOverloadsMessage" xml:space="preserve">
-    <value>{0} uses unsafe overload of '{1}' method</value>
+    <value>Uses unsafe overload of '{0}' method</value>
```
Comment (srivatsn) [inline src/Desktop.Analyzers/Core/DesktopAnalyzersResources.resx]: Why this change? Starting with uses sounds pretty blunt and without the context, there would be no way of distinguishing multiple instances of this error in the error list.

File: src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs [hunk 2]
```diff
@@ -31,85 +32,122 @@ public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
...
+
         public override void Initialize(AnalysisContext analysisContext)
         {
             analysisContext.RegisterCompilationStartAction(
                 (context) =>
                 {
                     Compilation compilation = context.Compilation;
-                    var xmlTypes = new CompilationSecurityTypes(compilation);
-                    if (ReferencesAnyTargetType(xmlTypes))
+                    var _xmlTypes = new CompilationSecurityTypes(compilation);
```
Comment (srivatsn) [inline src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs]: Only private fields are prefixed with _.

File: src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs [hunk 3]
```diff
@@ -31,85 +32,122 @@ public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
...
-                || types.XmlTextReader != null
-                || types.XPathDocument != null
-                || types.XmlSchema != null
-                || types.DataSet != null
-                || types.DataTable != null
-                || types.DataViewManager != null
-                || types.XmlSerializer != null;
-        }
+            #region Environment classes
+            private class XmlDocumentEnvironment
```
Comment (srivatsn) [inline src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs]: Should these be structs with readonly fields?

File: src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs [hunk 4]
```diff
@@ -367,475 +326,520 @@ private void AnalyzeNodeForDtdProcessingOverloads(SyntaxNodeAnalysisContext cont
...
+            {
+                IObjectCreationExpression objCreation = value as IObjectCreationExpression;
+
+                if (objCreation == null)
                 {
                     return;
                 }
 
-                if (property.MatchPropertyDerivedByName(_xmlTypes.XmlDocument, SecurityMemberNames.InnerXml))                                       //FxCop CA3058
+                if (_objectCreationOperationsAnalyzed.Contains(objCreation))
```
Comment (srivatsn) [inline src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs]: I don't understand this - why do you need to track the analyzed objectcreation expresiions?

Comment (qinxgit) [inline src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs]: ObjectCreation can be in a field declaration, can be in a variable declaration, or in a temporary object. If this is a variable declaration, this function will end up get invoked by both VariableDeclarationExpression and ObjectCreationExpression operation even though 1 is child of the other. this function this is just making sure we haven't seen this expression from somewhere else.

File: src/Desktop.Analyzers/Core/DesktopAnalyzersResources.resx (hunk 1 above)
Comment (qinxgit) [inline src/Desktop.Analyzers/Core/DesktopAnalyzersResources.resx]: We had an Email and I have you confirmed that we need to remove the first parameter messages as we using IOperation. It is no use since there is accurate location info in the issue.

File: src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs (hunk 2 above)
Comment (qinxgit) [inline src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs]: Will fix this.

File: src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs (hunk 3 above)
Comment (qinxgit) [inline src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs]: No. If structs, the object of the struct will get copied, instead of referenced by, when you pass it around thru functions, or put into dictionaries or get back from dictionaries. I definitely should avoid that.

File: src/Desktop.Analyzers/Core/DesktopAnalyzersResources.resx (hunk 1 above)
Comment (qinxgit) [inline src/Desktop.Analyzers/Core/DesktopAnalyzersResources.resx]: But I can update it to "Unsafe overload of '{1}' method"

File: src/Desktop.Analyzers/Core/DesktopAnalyzersResources.resx [hunk 5]
```diff
@@ -376,7 +376,7 @@
     <comment>CA3075</comment>
   </data>
   <data name="DoNotUseSetInnerXmlMessage" xml:space="preserve">
-    <value>{0} uses the unsafe setter of InnerXml property of System.Xml.XmlDocument.</value>
+    <value>Uses the unsafe setter of InnerXml property of System.Xml.XmlDocument.</value>
```
Comment (srivatsn) [inline src/Desktop.Analyzers/Core/DesktopAnalyzersResources.resx]: The tone of the messages are inconsistent - some point at a problem, some are more active an say "uses x' and some are not even complete sentences. 

@lgolding - can you please provide suggestions for the messages? If you want to tackle this as part of an all-up review of messages that you were planning, then we don't need to block Rick.

File: src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs (hunk 3 above)
Comment (qinxgit) [inline src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs]: Agree.

File: src/Desktop.Analyzers/Core/DesktopAnalyzersResources.resx (hunk 5 above)
Comment (lgolding) [inline src/Desktop.Analyzers/Core/DesktopAnalyzersResources.resx]: Yes, I will do this as part of the scrub. No need to block.

File: src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs [hunk 6]
```diff
@@ -367,475 +326,520 @@ private void AnalyzeNodeForDtdProcessingOverloads(SyntaxNodeAnalysisContext cont
...
-            private class XmlReaderSettingsEnvironment
+            private void AnalyzeObjectCreationOperation(OperationAnalysisContext context)
             {
-                internal SyntaxNode XmlReaderSettingsDefinition;  // REVIEW: remove?
-                internal ISymbol EnclosingConstructSymbol;        // REVIEW: remove?
-                internal bool IsDtdProcessingDisabled;
-                internal bool IsMaxCharactersFromEntitiesLimited;
-                internal bool IsSecureResolver;
-                internal readonly bool IsConstructedInCodeBlock;
+                AnalyzeObjectCreationInternal(context, null, context.Operation);
```
Comment (basoundr) [inline src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs]: Remove blank spaces at the end of the line

Comment (qinxgit) [inline src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs]: Done.

File: src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs [hunk 7]
```diff
@@ -137,215 +176,135 @@ public void AnalyzeCodeBlockEnd(CodeBlockAnalysisContext context)
...
+                        break;
+                }
             }
 
-            private void AnalyzeNodeForDtdProcessingOverloads(SyntaxNodeAnalysisContext context)
+            private void AnalyzeInvocation(OperationAnalysisContext context)
             {
-                SyntaxNode node = context.Node;
-                SemanticModel model = context.SemanticModel;
+                IInvocationExpression invoke = context.Operation as IInvocationExpression;
```
Comment (basoundr) [inline src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs]: Rename to `invocationExpression` instead of `invoke`. With so much code, variable naming helps a lot

File: src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs [hunk 8]
```diff
@@ -137,215 +176,135 @@ public void AnalyzeCodeBlockEnd(CodeBlockAnalysisContext context)
...
+                {
+                    return;
+                }
+
+                AnalyzeMethodOverloads(context, method, invoke.ArgumentsInParameterOrder, invoke.Syntax.GetLocation());
+            }
+
+            private void AnalyzeMethodOverloads(OperationAnalysisContext context, IMethodSymbol method, ImmutableArray<IArgument> arguments, Location location)
+            {
+                SemanticModel model = context.Compilation.GetSemanticModel(context.Operation.Syntax.SyntaxTree);
```
Comment (basoundr) [inline src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs]: the SemanticModel is not used until later. You can push the statement right before you start using the semantic model

File: src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs [hunk 9]
```diff
@@ -137,215 +176,135 @@ public void AnalyzeCodeBlockEnd(CodeBlockAnalysisContext context)
...
+
+                if (method.MatchMethodDerivedByName(_xmlTypes.XmlDocument, SecurityMemberNames.Load) ||                                    //FxCop CA3056
```
Comment (basoundr) [inline src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs]: I looked at all the callers of `MatchMethodDerivedByName` and it is always a IMethodSymbol. So if you change the method signature to `public static bool MatchMethodDerivedByName(this IMethodSymbol member, INamedTypeSymbol type, string name)` instead of `public static bool MatchMethodDerivedByName(this ISymbol member, INamedTypeSymbol type, string name)` then that method does not need the `member.Kind == SymbolKind.Method`

File: src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs [hunk 10]
```diff
@@ -137,215 +176,135 @@ public void AnalyzeCodeBlockEnd(CodeBlockAnalysisContext context)
...
+
+                if(method == null)
+                {
+                    return;
+                }
+
+                AnalyzeMethodOverloads(context, method, invoke.ArgumentsInParameterOrder, invoke.Syntax.GetLocation());
+            }
+
+            private void AnalyzeMethodOverloads(OperationAnalysisContext context, IMethodSymbol method, ImmutableArray<IArgument> arguments, Location location)
```
Comment (basoundr) [inline src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs]: `AnalyzeMethodOverloads` could be `private void AnalyzeMethodOverloads(OperationAnalysisContext context, IMethodSymbol method, IHasArgumentExpression expression)` and arguments & location can be obtained in appropriate places instead of having it computer as early

File: src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs (hunk 8 above)
Comment (qinxgit) [inline src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs]: Sure. Is it because the GetSemanticModel is costy?

File: src/Desktop.Analyzers/Core/Helpers/SecurityDiagnosticHelpers.cs [hunk 11]
```diff
@@ -164,6 +165,30 @@ public static int HasXmlReaderParameter(IMethodSymbol method, CompilationSecurit
             return HasSpecifiedParameter(method, xmlTypes, IsXmlReaderType);
```
Comment (basoundr) [inline src/Desktop.Analyzers/Core/Helpers/SecurityDiagnosticHelpers.cs]: `HasSpecifiedParameter` returns an `int`. Rename `HasSpecifiedParameter` to `GetSpecifiedParameterIndex`

File: src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs (hunk 10 above)
Comment (qinxgit) [inline src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs]: Note here the location is usually different from the location of the arguments expression. The caller will have a location to report on if there are issues found, which is different from the arguments expression.

File: src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs [hunk 12]
```diff
@@ -367,475 +326,520 @@ private void AnalyzeNodeForDtdProcessingOverloads(SyntaxNodeAnalysisContext cont
...
+                XmlDocumentEnvironment env = null;
 
-                if (lhs == null || rhs == null)
+                if (variable == null || !_xmlDocumentEnvironments.ContainsKey(variable))
                 {
-                    return;
+                    env = new XmlDocumentEnvironment();
+                    env.IsSecureResolver = false;
+                    env.IsXmlResolverSet = false;
+                    env.XmlDocumentDefinition = objCreation.Syntax;
```
Comment (basoundr) [inline src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs]: `AnalyzeObjectCreationForXmlDocument` is the same in both `if` and `else` case. Can move this statement out

File: src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs [hunk 13]
```diff
@@ -367,475 +326,520 @@ private void AnalyzeNodeForDtdProcessingOverloads(SyntaxNodeAnalysisContext cont
...
                 {
-                    XmlDocumentEnvironment env = new XmlDocumentEnvironment();
+                    var prop = init as IPropertyInitializer;
 
-                    if (rhsMethodSymbol.ContainingType != xmlTypes.XmlDocument)
+                    if (prop != null)
                     {
-                        env.IsSecureResolver = true;
-                    }
+                        if (prop.InitializedProperty.MatchPropertyDerivedByName(_xmlTypes.XmlDocument, "XmlResolver"))
```
Comment (basoundr) [inline src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs]: Method `MatchPropertyDerivedByName` does not need the check `member.Kind == SymbolKind.Property `

File: src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs [hunk 14]
```diff
@@ -367,475 +326,520 @@ private void AnalyzeNodeForDtdProcessingOverloads(SyntaxNodeAnalysisContext cont
...
-                    XmlDocumentEnvironment env = null;
-                    _xmlDocumentEnvironments.TryGetValue(lhsExpressionSymbol, out env);
-
-                    ITypeSymbol rhsType = model.GetTypeInfo(rhs).Type;
-                    // if XmlDocument was constructed in the same code block with default values.
-                    if (env != null)
-                    {
-                        env.IsXmlResolverSet = true;
-                    }
+                if (env != null)
```
Comment (basoundr) [inline src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs]: `env` can never be null

File: src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs (hunk 8 above)
Comment (basoundr) [inline src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs]: Everything inside a SemanticModel is computed lazily so it is just a shell and it wont be costly. But if we could avoid making the call in the first place then why do it.

File: src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs (hunk 10 above)
Comment (basoundr) [inline src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs]: The `location` that gets passed here is `invoke.Syntax.GetLocation()` and `objCreation.Syntax.GetLocation()` which can be obtained from the `expression.Syntax.GetLocation()` rather than having to call `GetLocation()` in 2 different places

File: src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs [hunk 15]
```diff
@@ -367,475 +326,520 @@ private void AnalyzeNodeForDtdProcessingOverloads(SyntaxNodeAnalysisContext cont
...
-                        }
+            private void AnalyzeObjectCreationForXmlReaderSettings(OperationAnalysisContext context, ISymbol variable, IObjectCreationExpression objCreation)
+            {
+                XmlReaderSettingsEnvironment xmlReaderSettingsEnv = new XmlReaderSettingsEnvironment(_isFrameworkSecure);
 
-                        XmlTextReaderEnvironment env = null;
-                        _xmlTextReaderEnvironments.TryGetValue(lhsExpressionSymbol, out env);
+                if (variable != null)
+                {
+                    _xmlReaderSettingsEnvironments[variable] = xmlReaderSettingsEnv;
```
Comment (basoundr) [inline src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs]: is it a guarantee that `variable` will always be present in `_xmlReaderSettingsEnvironments`?

File: src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs [hunk 16]
```diff
@@ -367,475 +326,520 @@ private void AnalyzeNodeForDtdProcessingOverloads(SyntaxNodeAnalysisContext cont
...
+            private void AnalyzeObjectCreationForXmlDocument(OperationAnalysisContext context, ISymbol variable, IObjectCreationExpression objCreation)
             {
-                SyntaxNode node = context.Node;
-                SemanticModel model = context.SemanticModel;
-
-                node = _syntaxNodeHelper.GetVariableDeclaratorOfAFieldDeclarationNode(node) ?? node;
-
-                SyntaxNode lhs = _syntaxNodeHelper.GetAssignmentLeftNode(node);
-                SyntaxNode rhs = _syntaxNodeHelper.GetAssignmentRightNode(node);
+                XmlDocumentEnvironment env = null;
```
Comment (basoundr) [inline src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs]: Rename `env` to `XmlDocumentEnvironment `

File: src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs [hunk 17]
```diff
@@ -367,475 +326,520 @@ private void AnalyzeNodeForDtdProcessingOverloads(SyntaxNodeAnalysisContext cont
...
+
+            private void AnalyzeVariableDeclaration(OperationAnalysisContext context)
+            {
+                IVariableDeclaration declare = context.Operation as IVariableDeclaration;
+                AnalyzeObjectCreationInternal(context, declare.Variable, declare.InitialValue);
+            }
+
+            private void AnalyzeXmlResolverPropertyAssignmentForXmlDocument(OperationAnalysisContext context, ISymbol assignedSymbol, IAssignmentExpression expression)
+            {
+                XmlDocumentEnvironment env = null;
```
Comment (basoundr) [inline src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs]: rename `env` to `xmlDocumentEnv`?

File: src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs [hunk 18]
```diff
@@ -367,475 +326,520 @@ private void AnalyzeNodeForDtdProcessingOverloads(SyntaxNodeAnalysisContext cont
...
+                bool isSecureResolver = false;
+
+                if (_xmlDocumentEnvironments.ContainsKey(assignedSymbol))
```
Comment (basoundr) [inline src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs]: This `if` block can moved to the bottom of the method call and be merged with `if (env != null)`

File: src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs [hunk 19]
```diff
@@ -367,475 +326,520 @@ private void AnalyzeNodeForDtdProcessingOverloads(SyntaxNodeAnalysisContext cont
...
             }
 
-            //Note: False negative if integer is used to set DtdProcessing instead of enumeration
-            private void AnalyzeNodeForXmlReaderSettings(SyntaxNodeAnalysisContext context)
+            private void AnalyzeXmlTextReaderProperties(OperationAnalysisContext context, ISymbol assignedSymbol, IAssignmentExpression expression, bool isXmlTextReaderXmlResolverProperty, bool isXmlTextReaderDtdProcessingProperty)
             {
-                SyntaxNode node = context.Node;
-                SemanticModel model = context.SemanticModel;
+                XmlTextReaderEnvironment env = null;
+                _xmlTextReaderEnvironments.TryGetValue(assignedSymbol, out env);
```
Comment (basoundr) [inline src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs]:
```C#
_xmlTextReaderEnvironments.TryGetValue(assignedSymbol, out env);

if (env == null)
{
    env = new XmlTextReaderEnvironment(_isFrameworkSecure);
}
```

can be replaced with 
```#
XmlTextReaderEnvironment env = null;
if(!_xmlTextReaderEnvironments.TryGetValue(assignedSymbol, out env))
{
    env = new XmlTextReaderEnvironment(_isFrameworkSecure);
}
```

File: src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs [hunk 20]
```diff
@@ -367,475 +326,520 @@ private void AnalyzeNodeForDtdProcessingOverloads(SyntaxNodeAnalysisContext cont
...
+                }
+            }
 
-                    env.XmlReaderSettingsDefinition = node;
-                    env.EnclosingConstructSymbol = _syntaxNodeHelper.GetEnclosingConstructSymbol(node, model);
+            private void AnalyzeAssignment(OperationAnalysisContext context)
+            {
+                IAssignmentExpression expression = context.Operation as IAssignmentExpression;
+                
+                SemanticModel model = context.Compilation.GetSemanticModel(expression.Syntax.SyntaxTree);
```
Comment (basoundr) [inline src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs]: move this statement after the `if` statement

File: src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs [hunk 21]
```diff
@@ -137,215 +176,135 @@ public void AnalyzeCodeBlockEnd(CodeBlockAnalysisContext context)
...
-                        );
+                        context.ReportDiagnostic(diag);
                     }
                     else
                     {
-                        SyntaxNode settingsNode = _syntaxNodeHelper.GetInvocationArgumentExpressionNodes(node).ElementAt(xmlReaderSettingsIndex);
-                        ISymbol settingsSymbol = SyntaxNodeHelper.GetSymbol(settingsNode, model);
+
+                        IArgument arg = arguments[xmlReaderSettingsIndex];
+                        ISymbol settingsSymbol = arg.Value.Syntax.GetDeclaredOrReferencedSymbol(model);
```
Comment (basoundr) [inline src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs]: I would make `GetDeclaredOrReferencedSymbol` an extension of `SemanticModel` rather than `SyntaxNode` since the calls inside `GetDeclaredOrReferencedSymbol` are on `SemanticModel` rather than `SyntaxNode`

**Comment by basoundr** [comment]
There are also quite a few helper methods that are no longer needed in `SecurityDiagnosticHelpers` like `IsXmlDocumentXmlResolverPropertyDerived`,etc

File: src/Desktop.Analyzers/Core/Helpers/SecurityDiagnosticHelpers.cs [hunk 22]
```diff
@@ -164,6 +165,30 @@ public static int HasXmlReaderParameter(IMethodSymbol method, CompilationSecurit
...
+        public static bool IsExpressionEqualsNull(IOperation operation)
+        {
+            ILiteralExpression literal = operation as ILiteralExpression;
+            return literal != null && literal.ConstantValue.HasValue && literal.ConstantValue.Value == null; 
+        }
+
+        public static bool IsExpressionEqualsDtdProcessingParse(IOperation operation)
+        {
+            IFieldReferenceExpression enumRef = operation as IFieldReferenceExpression;
+            return enumRef != null && enumRef.ConstantValue.HasValue && (int)enumRef.ConstantValue.Value == 2; // DtdProcessing.Parse
```
Comment (basoundr) [inline src/Desktop.Analyzers/Core/Helpers/SecurityDiagnosticHelpers.cs]: You have the `DtdProcessing` type under SecurityTypes. Can you verify if the value refers to the Field Symbol `DtdProcessing.Parse`? I am not really sure how feasible that is here

File: src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs [hunk 23]
```diff
@@ -15,15 +15,16 @@ namespace Desktop.Analyzers
     /// <summary>
     /// Secure DTD processing and entity resolution in XML
     /// </summary>
-    public abstract class DoNotUseInsecureDTDProcessingAnalyzer<TLanguageKindEnum> : DiagnosticAnalyzer where TLanguageKindEnum : struct
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
```
Comment (basoundr) [inline src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs]: +1 :)

File: src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs (hunk 10 above)
Comment (qinxgit) [inline src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs]: I now see your point. Will update it.

File: src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs (hunk 9 above)
Comment (qinxgit) [inline src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs]: Sure.

File: src/Desktop.Analyzers/Core/Helpers/SecurityDiagnosticHelpers.cs (hunk 11 above)
Comment (qinxgit) [inline src/Desktop.Analyzers/Core/Helpers/SecurityDiagnosticHelpers.cs]: Sure.

File: src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs [hunk 24]
```diff
@@ -137,215 +176,135 @@ public void AnalyzeCodeBlockEnd(CodeBlockAnalysisContext context)
...
+                    method.MatchMethodDerivedByName(_xmlTypes.XmlSchema, SecurityMemberNames.Read) ||                                      //FxCop CA3060
+                    method.MatchMethodDerivedByName(_xmlTypes.DataSet, SecurityMemberNames.ReadXml) ||                                     //FxCop CA3063
+                    method.MatchMethodDerivedByName(_xmlTypes.DataSet, SecurityMemberNames.ReadXmlSchema) ||                               //FxCop CA3064
+                    method.MatchMethodDerivedByName(_xmlTypes.XmlSerializer, SecurityMemberNames.Deserialize) ||                           //FxCop CA3070
+                    method.MatchMethodDerivedByName(_xmlTypes.DataTable, SecurityMemberNames.ReadXml) ||                                   //FxCop CA3071
+                    method.MatchMethodDerivedByName(_xmlTypes.DataTable, SecurityMemberNames.ReadXmlSchema))                               //FxCop CA3072
                 {
-                    if (SecurityDiagnosticHelpers.HasXmlReaderParameter(method, xmlTypes) < 0)
+                    if (SecurityDiagnosticHelpers.HasXmlReaderParameter(method, _xmlTypes) < 0)
```
Comment (qinxgit) [inline src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs]: Sure.

File: src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs (hunk 12 above)
Comment (qinxgit) [inline src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs]: Sure.

File: src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs (hunk 13 above)
Comment (qinxgit) [inline src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs]: Why? No one has checked it is a property or not. If a method has same name then this method named "Match**Property**DerivedByName" will return true.
Here is body 
``return member != null && member.Kind == SymbolKind.Property && member.MatchMemberDerivedByName(type, name);``

Here is body of MatchMemberDerivedByName
``
public static bool MatchMemberDerivedByName(this ISymbol member, INamedTypeSymbol type, string name)
        {
            return member != null && member.ContainingType.DerivesFrom(type) && member.MetadataName == name;
        }
``

File: src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs (hunk 14 above)
Comment (qinxgit) [inline src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs]: Sure.

File: src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs (hunk 15 above)
Comment (qinxgit) [inline src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs]: Well, it should be, if a XmlReaderSetting is created or declared, and captured by this analyzer, it should have an item in `_xmlReaderSettingsEnvironments`. If it is just a method parameter, no. I am not sure I answered the question.

File: src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs (hunk 16 above)
Comment (qinxgit) [inline src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs]: Sure.

File: src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs (hunk 17 above)
Comment (qinxgit) [inline src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs]: Sure.

File: src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs (hunk 18 above)
Comment (qinxgit) [inline src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs]: Sure.

File: src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs (hunk 19 above)
Comment (qinxgit) [inline src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs]: Sure.

File: src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs (hunk 20 above)
Comment (qinxgit) [inline src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs]: Sure.

File: src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs (hunk 21 above)
Comment (qinxgit) [inline src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs]: The intention of this extension method is to find what symbol is being declared or referenced within this SyntaxNode node. model is only a helper class here to retrieve symbol information. So  with that said, it is more specific to this node, instead of the entire model. When we look at where a extension method belongs to, we should look at the intention of this method instead of implementation. So I think it is better to stay with SyntaxNode..

File: src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs (hunk 23 above)
Comment (qinxgit) [inline src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs]: Removed all methods with no references.

File: src/Desktop.Analyzers/Core/Helpers/SecurityDiagnosticHelpers.cs (hunk 22 above)
Comment (qinxgit) [inline src/Desktop.Analyzers/Core/Helpers/SecurityDiagnosticHelpers.cs]: Yes, I have test cases. The reason I put 2 here is because in the current .NET CLR verion, this enum value is gone. To check it for other versions, a const will do the job.

**Comment by qinxgit** [comment]
@basoundr @srivatsn 
Is it good to sign off now?

**Comment by basoundr** [comment]
I was looking at the helpers for Propertyties and they too can be modified to accept IPropertySymbol as the parameter instead of ISymbol. Pending that, this looks good +1

**Comment by qinxgit** [comment]
@basoundr   Can you show me where I need to change?

**Comment by basoundr** [comment]
`public static bool IsXmlDocumentXmlResolverProperty(ISymbol symbol, CompilationSecurityTypes xmlTypes)`, `public static bool IsXmlTextReaderXmlResolverProperty(ISymbol symbol, CompilationSecurityTypes xmlTypes)`and many more.

The above mentioned are the callers of `private static bool IsSpecifiedProperty(ISymbol symbol, INamedTypeSymbol namedType, string propertyName)` and `private static bool IsSpecifiedPropertyDerived(ISymbol symbol, INamedTypeSymbol namedType, string propertyName)
`.

The helpers if modified as `private static bool IsSpecifiedProperty(IPropertySymbol symbol, INamedTypeSymbol namedType, string propertyName)` and `private static bool IsSpecifiedPropertyDerived(IPropertySymbol symbol, INamedTypeSymbol namedType, string propertyName)` and the corresponding callers are modified then things are good :)

**Comment by qinxgit** [comment]
@basoundr The ISymbol are updated to IPropertySymbol

**Comment by qinxgit** [comment]
@basoundr Good to go?

**Comment by basoundr** [comment]
Thanks @qinxgit :+1:
